This patch was developed in-house, contributed upstream, and accepted with
minor modifications for 8.15.2; when that release comes out, it should no
longer be needed.

--- sendmail-8.15.1/RELEASE_NOTES.~1~	2015-03-02 10:53:48.866427766 -0800
+++ sendmail-8.15.1/RELEASE_NOTES	2015-03-02 10:55:13.925190180 -0800
@@ -5,6 +5,13 @@
 of the sendmail configuration files, the date of release, and a
 summary of the changes in that release.
 
+	CONFIG: New option UseCompressedIPv6Addresses to select between
+		compressed and uncompressed IPv6 addresses.  The default
+		value depends on the compile-time option IPV6_FULL:
+		For 1 the default is False, for 0 it is True, thus
+		preserving the current behaviour.  Based on patch from
+		John Beck of Oracle.
+
 8.15.1/8.15.1	2014/12/06
 	SECURITY: Properly set the close-on-exec flag for file descriptors
 		(except stdin, stdout, and stderr) before executing mailers.
--- sendmail-8.15.1/cf/README.~2~	2015-03-02 10:56:39.888131319 -0800
+++ sendmail-8.15.1/cf/README	2015-03-02 10:56:39.926065186 -0800
@@ -3454,6 +3454,12 @@
 					cached but otherwise idle connection
 					to a host will prevent other sendmails
 					from connecting to the other host.
+confUSE_COMPRESSED_IPV6_ADDRESSES
+			UseCompressedIPv6Addresses
+					[undefined] If set, use the compressed
+					form of IPv6 addresses, such as
+					IPV6:::1, instead of the uncompressed
+					form, such as IPv6:0:0:0:0:0:0:0:1.
 confUSE_ERRORS_TO*	UseErrorsTo	[False] Use the Errors-To: header to
 					deliver error messages.  This should
 					not be necessary because of general
--- sendmail-8.15.1/cf/m4/proto.m4.~2~	2015-03-02 10:59:56.762362990 -0800
+++ sendmail-8.15.1/cf/m4/proto.m4	2015-03-02 10:59:56.821388655 -0800
@@ -329,6 +329,9 @@
 # use Errors-To: header?
 _OPTION(UseErrorsTo, `confUSE_ERRORS_TO', `False')
 
+# use compressed IPv6 address format?
+_OPTION(UseCompressedIPv6Addresses, `confUSE_COMPRESSED_IPV6_ADDRESSES', `')
+
 # log level
 _OPTION(LogLevel, `confLOG_LEVEL', `10')
 
--- sendmail-8.15.1/doc/op/op.me.~1~
+++ sendmail-8.15.1/doc/op/op.me
@@ -8360,6 +8360,12 @@ are always unsafe.
 Note: use
 .b DontBlameSendmail
 instead; this option is deprecated.
+.ip UseCompressedIPv6Addresses
+[no short name]
+If set, the compressed format of IPv6 addresses,
+such as IPV6:::1, will be used,
+instead of the uncompressed format,
+such as IPv6:0:0:0:0:0:0:0:1.
 .ip UseErrorsTo
 [l]
 If there is an
--- sendmail-8.15.1/sendmail/conf.c.~1~	2014-11-07 10:01:17.000000000 -0800
+++ sendmail-8.15.1/sendmail/conf.c	2015-03-02 11:02:56.829822173 -0800
@@ -379,6 +379,9 @@
 	}
 	else
 		InetMode = AF_INET;
+#if !IPV6_FULL
+	UseCompressedIPv6Addresses = true;
+#endif
 #else /* NETINET6 */
 	InetMode = AF_INET;
 #endif /* NETINET6 */
@@ -5756,7 +5759,7 @@
 	"HES_GETMAILHOST",
 #endif
 #if IPV6_FULL
-	/* Use uncompressed IPv6 address format (no "::") */
+	/* Use uncompressed IPv6 address format (no "::") by default */
 	"IPV6_FULL",
 #endif
 #if LDAPMAP
--- a/sendmail/daemon.c
+++ b/sendmail/daemon.c
@@ -4261,12 +4261,10 @@ anynet_ntop(s6a, dst, dst_len)
 			return NULL;
 		dst += sz;
 		dst_len -= sz;
-# if IPV6_FULL
-		ap = sm_inet6_ntop(s6a, dst, dst_len);
-# else /* IPV6_FULL */
-		ap = (char *) inet_ntop(AF_INET6, s6a, dst, dst_len);
-# endif /* IPV6_FULL */
-
+		if (UseCompressedIPv6Addresses)
+			ap = (char *) inet_ntop(AF_INET6, s6a, dst, dst_len);
+		else
+			ap = sm_inet6_ntop(s6a, dst, dst_len);
 		/* Restore pointer to beginning of string */
 		if (ap != NULL)
 			ap = d;
--- sendmail-8.15.1/sendmail/readcf.c.~1~	2014-10-22 09:33:53.000000000 -0700
+++ sendmail-8.15.1/sendmail/readcf.c	2015-03-02 11:04:38.892714939 -0800
@@ -2473,6 +2473,8 @@
 # define O_ADDBCC 0xeb
 	{ "AddBcc",			O_ADDBCC,	OI_NONE },
 #endif
+#define O_USECOMPRESSEDIPV6ADDRESSES 0xec
+	{ "UseCompressedIPv6Addresses",	O_USECOMPRESSEDIPV6ADDRESSES, OI_NONE },
 
 	{ NULL,				'\0',		OI_NONE	}
 };
@@ -4277,6 +4279,9 @@
 		AddBcc = atobool(val);
 		break;
 #endif
+	  case O_USECOMPRESSEDIPV6ADDRESSES:
+		UseCompressedIPv6Addresses = atobool(val);
+		break;
 
 	  default:
 		if (tTd(37, 1))
--- sendmail-8.15.1/sendmail/sendmail.h.~1~	2014-11-07 05:26:39.000000000 -0800
+++ sendmail-8.15.1/sendmail/sendmail.h	2015-03-02 11:05:51.115921999 -0800
@@ -2384,6 +2384,7 @@
 EXTERN bool	WorkAroundBrokenAAAA;	/* some nameservers return SERVFAIL on AAAA queries */
 EXTERN bool	UseErrorsTo;	/* use Errors-To: header (back compat) */
 EXTERN bool	UseNameServer;	/* using DNS -- interpret h_errno & MX RRs */
+EXTERN bool	UseCompressedIPv6Addresses;	/* for more specific zero-subnet matches */
 EXTERN char	InetMode;		/* default network for daemon mode */
 EXTERN char	OpMode;		/* operation mode, see below */
 EXTERN char	SpaceSub;	/* substitution for <lwsp> */
